apps/src/music/progress/MusicValidator.ts

```diff
@@ -47,62 +49,97 @@ export default class MusicValidator extends Validator {
-        eventData.when + length > currentPlayheadPosition
-      ) {
-        currentNumberSounds++;
 
-        if (eventData.triggered) {
+      if (eventData.type === 'sound') {
+        if (
+          eventData.when <= currentPlayheadPosition &&
+          eventData.when + length > currentPlayheadPosition
+        ) {
+          currentNumberSounds++;
+
+          if (eventData.triggered) {
+            this.conditionsChecker.addSatisfiedCondition({
+              name: MusicConditions.PLAYED_SOUND_TRIGGERED.name,
+            });
+          }
+
           this.conditionsChecker.addSatisfiedCondition({
-            name: MusicConditions.PLAYED_SOUND_TRIGGERED.name,
+            name: MusicConditions.PLAYED_SOUND_ID.name,
+            value: eventData.id,
           });
         }
 
-        this.conditionsChecker.addSatisfiedCondition({
-          name: MusicConditions.PLAYED_SOUND_ID.name,
-          value: eventData.id,
-        });
-      }
-
-      if (eventData.when <= currentPlayheadPosition) {
-        playedNumberSounds++;
+        if (eventData.when <= currentPlayheadPosition) {
+          playedNumberSounds++;
+        }
+      } else if (eventData.type === 'pattern') {
+        if (eventData.when <= currentPlayheadPosition) {
+          playedNumberPatterns++;
+        }
+      } else if (eventData.type === 'chord') {
+        if (eventData.when <= currentPlayheadPosition) {
+          playedNumberChords++;
+        }
       }
     });
 
     // Check for up to a certain number of sounds playing simultaneously.
     // Note that if, for example, 3 sounds are playing, then we'll consider
     // that 2 sounds and 1 sound have also been played together.
-    const maxNumberSoundsTogether = 3;
     for (
-      let numberSounds = maxNumberSoundsTogether;
+      let numberSounds = currentNumberSounds;
       numberSounds >= 1;
       numberSounds--
     ) {
-      if (currentNumberSounds >= numberSounds) {
-        this.conditionsChecker.addSatisfiedCondition({
-          name: MusicConditions.PLAYED_SOUNDS_TOGETHER.name,
-          value: currentNumberSounds,
-        });
-      }
+      this.conditionsChecker.addSatisfiedCondition({
+        name: MusicConditions.PLAYED_SOUNDS_TOGETHER.name,
+        value: numberSounds,
+      });
     }
 
     // Check for up to a certain number of sounds played.
-    const maxNumberSounds = 3;
     for (
-      let numberSounds = maxNumberSounds;
+      let numberSounds = playedNumberSounds;
       numberSounds >= 1;
       numberSounds--
     ) {
-      if (playedNumberSounds >= numberSounds) {
-        this.conditionsChecker.addSatisfiedCondition({
-          name: MusicConditions.PLAYED_SOUNDS.name,
-          value: playedNumberSounds,
-        });
-      }
+      this.conditionsChecker.addSatisfiedCondition({
+        name: MusicConditions.PLAYED_SOUNDS.name,
+        value: numberSounds,
+      });
+    }
+
+    // Check for up to a certain number of patterns played.
+    for (
```

Comment:
worth creating a small helper for these? something like `addPlayedConditions(conditionName: string, playedNumber: number)`?

Reply:
Good idea; done.